### DISCUSSION TITLE
Alternative to using Next.js for SEO with react-native-web?

### DISCUSSION THREAD

**Discussion by mayank-budhiraja** [necolas/react-native-web#2192] [Help]
I need SEO for the web app but without next.js I can do it. Is there any other alternative? or just simply use a wrapper above screens for basic SEO

**Comment by cglacet** [comment]
Did you tried server side rendering? I think the question is related. Maybe some people here have experience with SSR?

I'm also very interested by this SEO question.

**Comment by Freddy03h** [comment]
I'm using [Razzle](https://razzlejs.org) for my RNW app

**Reply by MiguelNiblock** [reply]
The thing with Razzle is you have to set up route handling and serverside data loading yourself, which is a herculean task for a serious app. Nextjs does everything for you. Then there's Remix, which is in many ways similar to Nextjs, but the downside is you can't configure the esbuild compiler, so you can't specify things like module aliases, or transpiled modules. So after investigating much, I still think Nextjs is the best choice.

**Reply by Freddy03h** [reply]
Like each time, it depend of your product, let me explain my use case:

My main product is a react-native app, on production for Android and iOS, and working well on a tablet and offline first. And I wanted to create the PWA twin, but didn't want to use react-navigation for the web version, because of some limitations on navigation and UI logic: I want to do it differently on the web app.
I could use react-router but I've done a router on my own (not that hard, really), so it was easier to encapsulate navigation logic inside custom hooks used by views and components, so navigation work well for the native app and the web app.

Next.js is an other level of dependency in your product. First, it's a subjective opinion but I'm not a fan of the page logic, the routing based on folders. Also, the way to do data fetching will make so much impact on my existing app architecture (a complexity that the native app doesn't need), also I really don't need SSR for logged views.

The web app has a service worker for working offline, so the SSR is only for non-logged pages for SEO and SMO (good preview of shared link on social network), it has a small impact in user experience because it'll work only for the first visit (my product is a tool, it can be opened several times a week but has no long session, and need to work offline).

With theses constrains in mind, using Razzle was the best choice for me, because doing routing and data loading for some route is not an herculean task, and it'll be more easily removable when we'll not need to do SSR for SEO/SMO later.

I'm agree with you that Next.js is a default web solution for many products, but for my need, it's too tied to the web platform. That's why I wanted to share an other alternative like Razzle, because people interested in react-native-web are often interested on sharing code between react-native and web app without having big impact of a specific platform in all the code base and architecture of their app.

**Reply by MiguelNiblock** [reply]
Looking at it that way I agree, you could do server-side data loading only for the few screens that really need ssr/ssg which are a few public routes. For the rest, the authenticated routes, you will continue to do client side loading and rendering. 
@Freddy03h did you actually end up using the react-navigation container for SSR? https://reactnavigation.org/docs/server-rendering/

that is one of the main reasons why i was interested in razzle, using this instead of file-system based routing configuration.

**Reply by Freddy03h** [reply]
No I'm not using react-navigation on web.

(It's off topic, but I'm using rescript, so a variant type, pattern matching and the default small [Router](https://rescript-lang.org/docs/react/latest/router) in rescript-react is enough)

**Reply by flsilva** [reply]
@MiguelNiblock Remix looks awesome, it seems to simplify a lot of things out of the box and in a very nice way, but I agree it's not flexible regarding its bundle pipeline configuration, and it seems 100% focused on web apps (for now at least). It might take a while for it to start being used with React Native for serious apps, if at all, but I wish they take that path, supporting cross-platform and RNW apps seamlessly. If that was the case, I'd build my next RNW app on top of it for sure. But for now Next.js seems like a very good solution to me.

@Freddy03h I'm not a fan of the page logic and routing based on files and folders on Next.js either, it'd be much more flexible if it was based on simple JS configuration. So I can see why it makes sense for you to use something more flexible like Razzle, since you already had a React Native-only app with its own file and folder structure, and the web app came later. One important thing though: even though you're right that SSR will work only for the first visit, users might visit (open) your web app many times a day, several days a week, so that's a lot of "first visits", and adds up to a far from ideal user experience. Per what you mentioned, I'm assuming you're CSR (Client-Side Rendering) your app when a user logs in. That's what we've been doing for ages now in the SPA land, I did that a lot of times in the past too, but by doing that our apps usually take several seconds to bootstrap and render. I think it's time for us to improve that on every app we write.

I strongly recommend you to read [Remix vs Next.js](https://remix.run/blog/remix-vs-next) from Ryan Florence (Remix co-founder), to understand why he states that "...server rendering is about better performance, too, not just SEO.". When a logged in user visits a web app, the web app can make several requests right away to bootstrap a user's dashboard, etc. When you CSR it takes a lot more time to [FCP](https://developer.mozilla.org/en-US/docs/Glossary/First_contentful_paint) (First Contentful Paint) i and  [TTI](https://developer.mozilla.org/en-US/docs/Glossary/Time_to_interactive) (Time to Interactive) than when you SSR. Ryan does a great job explaining it in that post. Another huge benefit is that your code can be much simpler if SSR is your only rendering implementation (as is the case with Remix). It even replaces the need for SSG if you run on the edge, as he explains, and cache content in a CDN (for requests that makes sense caching, which might include dynamic requests like product searches that he explains). So you kind of get the best of both (three?) rendering worlds. That looks very promising and a superior solution than Next.js to me. But since cross-platform apps using RNW are very appealing to me as an engineer, saving a lot of time and money for my clients (those who of course benefit from an app deployed to those three platforms), I'm still using Next.js, but I'd love to give Remix a try when they make it possible without hacking their platform.

@mayank-budhiraja For web-only apps, Remix is probably better (per what I've read), but for cross-platform RNW apps I still think Next.js is the best and most flexible solution since Remix doesn't support RNW (at least not without some hacking and probably some other issues and limitations).

Let me know your thoughts.

**Reply by Freddy03h** [reply]
@flsilva 
> One important thing though: even though you're right that SSR will work only for the first visit, users might visit (open) your web app many times a day, several days a week, so that's a lot of "first visits", and adds up to a far from ideal user experience.

It's not how it work. As I mentioned the app works offline, it mean the app shell is stored locally, so when the user make a 2nd visit. The app shell is loaded from the cache, the SSR server is not requested. It work like a mobile app you installed on your device.

**Reply by abdulragib** [reply]
> @flsilva 
> > One important thing though: even though you're right that SSR will work only for the first visit, users might visit (open) your web app many times a day, several days a week, so that's a lot of "first visits", and adds up to a far from ideal user experience.
> 
> It's not how it work. As I mentioned the app works offline, it mean the app shell is stored locally, so when the user make a 2nd visit. The app shell is loaded from the cache, the SSR server is not requested. It work like a mobile app you installed on your device.
> 

Which local database you use for web and mobile

**Comment by getlocaladmin** [comment]
I suggest checking out Solito https://solito.dev/.
